pangeo_forge_recipes/serialization.py

```diff
@@ -0,0 +1,133 @@
+        default=json_default,
+        ensure_ascii=False,
+        sort_keys=True,
+        indent=None,
+        separators=(",", ":"),
+    )
+    return sha256(b.encode("utf-8")).digest()
+
+
+def dict_drop_empty(pairs: Sequence[Sequence]) -> dict:
+    """Custom factory function for ``dataclasses.asdict`` which drops fields for which the value is
+    ``None`` or for which the value is an empty collection (e.g. an empty ``list`` or ``dict``).
+
+    :param pairs: A sequence (list or tuple) of sequences of length 2, in which the first element of
+      each inner sequence is a hashable which can serve as a dictionary key, and the second element
+      of each inner sequence is the value to map to the key.
+    """
+
+    # https://death.andgravity.com/stable-hashing#problem-we-need-to-ignore-empty-values
+
+    return dict((k, v) for k, v in pairs if not (v is None or not v and isinstance(v, Collection)))
+
+
+def dataclass_sha256(dclass: type, ignore_keys: List[str]) -> bytes:
+    """Generate a deterministic sha256 hash from a Python ``dataclass``. Fields for which the value
+    is either ``None`` or an empty collection are excluded from the hash calculation automatically.
+    To manually exclude other fields from the calculation, pass their names via ``igonore_keys``.
+    For field values which are not json serializable, type-specific handlers are defined by the
+    ``json_default`` function in this module.
+
+    :param dclass: The dataclass for which to calculate a hash.
+    :param ignore_keys: A list of field names to exclude from the hash calculation.
+    """
+
+    d = asdict(dclass, dict_factory=dict_drop_empty)
+    for k in ignore_keys:
+        del d[k]
+    return dict_to_sha256(d)
+
+
+def pattern_blockchain(pattern: FilePattern) -> List[bytes]:
+    """For a ``FilePattern`` instance, compute a blockchain, i.e. a list of hashes of length N+1,
+    where N is the number of index:filepath pairs yielded by the ``FilePattern`` instance's
+    ``.items()`` method. The first item in the list is calculated by hashing instance attributes.
+    Each subsequent item is calculated by hashing the byte string produced by concatenating the next
+    index:filepath pair yielded by ``.items()`` with the previous hash in the list.
+
+    :param pattern: The ``FilePattern`` instance for which to calculate a blockchain.
+    """
+
+    # we exclude the format function and combine dims from ``root`` because they determine the
+    # index:filepath pairs yielded by iterating over ``.items()``. if these pairs are generated in
+    # a different way in the future, we ultimately don't care.
+    root = {
+        "fsspec_open_kwargs": pattern.fsspec_open_kwargs,
+        "query_string_secrets": pattern.query_string_secrets,
+        "file_type": pattern.file_type,
+        "nitems_per_file": [
+            op.nitems_per_file  # type: ignore
+            for op in pattern.combine_dims
+            if op.name in pattern.concat_dims
+        ],
```

Comment:
Should we make the value here a dict instead of a list? `op.name: op.nitems_per_file`

Reply:
yes I think this is the last thing will do that now

Reply:
👍 https://github.com/pangeo-forge/pangeo-forge-recipes/pull/349/commits/b48849c8bdbfa0e96ed75432fb476dd779dc6337